### PR TITLE
Enable use of RuntimeVariant to exclude runtime tests

### LINKED ISSUE #49653: Exclude Runtime Tests based on Runtime Mode [Interpreter/JIT/AOT]
We currently have the capability to exclude runtime tests using the issues.targets file, based on operating system, architecture, and runtime flavor (coreclr/mono).

There are some tests that fail on code generation or other mode of one of the runtimes, for example, mono-intepreter. Currently, to exclude them, we need to exclude the test for all runs on the runtime flavor.

This is a proposal to extend the issues.targets mechanism to handle these cases, to exclude tests only on particular configurations of the RuntimeFlavor.

### PR DESCRIPTION
**Pull request by naricc**
Currently, we can only exclude tests based on architecture, runtime flavor (mono/coreclr), and operating system. This change additionally allows the exclusions of tests based on runtime variants, such as interpreter, jit, or llvmaot.

I haven't gotten a full green build out of this yet, but I believe it trying to run the right things now. So I am making it ready for review.

Fixes https://github.com/dotnet/runtime/issues/49653

### CODE REVIEW

**Comment by SamMonoRT** [comment]
@naricc - while thinking about this, also consider the possibility of adding an extra refinement for differentiating Interpreter WASM  v/s AOT WASM test runs. Quite likely the set of failures maybe different.

File: src/tests/issues.targets [hunk 1]
```diff
@@ -1937,10 +1937,11 @@
         <ExcludeList Include = "$(XunitTestBinBase)tracing/eventsource/eventsourcetrace/eventsourcetrace/**">
             <Issue>needs triage</Issue>
         </ExcludeList>
+    </ItemGroup>
 
-        <!-- The following only fail in the mono interpreter, but we don't have a way to exclude based on scenario yet -->
-
-        <ExcludeList Include = "$(XunitTestBinBase)/JIT/jit64/opt/cse/HugeArray1/**">
+    <!-- Known failures for mono runtime on *all* architectures/operating systems in interpreter runtime mode -->
```
Comment (lambdageek) [inline src/tests/issues.targets:1942]: 🎉

**Comment by naricc** [comment]
@SamMonoRT Yeah, I had considered that. For interpreter wasm we can already just look at the architecture and variant; same for aot. The mechanism is general enough that we should be able to use it for any of the jit/aot/aotllvm things we have, just by defining a new variant for each.

It would be nice if I could create an enum for this, but there is no such facility in msbuild, or any way to statically check that the names agree with what it is in the build yml files.

**Comment by radical** [comment]
Can we surface this in `PlatformDetection` (or similar) so that it can be used with attributes like `ActiveIssue`? You will probably have to use `#ifdef FOO`+msbuild to define `FOO`. It would be useful [here](https://github.com/dotnet/runtime/pull/51724).

**Comment by radical** [comment]
I was trying https://gist.github.com/radical/dcbadb80909bd78fbabc4020a6018298#file-foo-diff-L74-L96, but that wouldn't be enough.

**Comment by naricc** [comment]
@radical Maybe. I am not sure what you had in mind exactly, adding a "GetRuntimeVariant" method to PlatformDetection?

Currently this is only for the runtime tests, which have their own unique way of building and don't use PlatfromDetection.

**Comment by radical** [comment]
> @radical Maybe. I am not sure what you had in mind exactly, adding a "GetRuntimeVariant" method to PlatformDetection?
> 
> Currently this is only for the runtime tests, which have their own unique way of building and don't use PlatfromDetection.

Oh ok, I was mistaken then. I'll add something for the library tests.

File: src/tests/issues.targets [hunk 2]
```diff
@@ -2668,12 +2670,14 @@
         <ExcludeList Include = "$(XunitTestBinBase)/JIT/Regression/VS-ia64-JIT/M00/b108366/b108366/**">
             <Issue>needs triage</Issue>
         </ExcludeList>
+    </ItemGroup>
 
+     <!-- These are known failures on mono-arm64 in the interpreter runtime variant -->
+    <ItemGroup Condition=" '$(RuntimeFlavor)' == 'mono' and ('$(TargetArchitecture)' == 'arm64' or '$(AltJitArch)' == 'arm64')  and '$(TargetsWindows)' != 'true' and '$(RuntimeVariant)' == 'monointerpreter' " >
```
[the remark this answers is not in the document]
Comment (naricc) [reply src/tests/issues.targets:2676]: Its is a property used by the coreclr cross compiler, although I don't understand exactly what it is for. I was parroting the condition that was already here, but we probably don't need it for mono.

File: eng/pipelines/common/templates/runtimes/run-test-job.yml [hunk 3]
```diff
@@ -63,7 +63,11 @@ jobs:
       - ${{ if notIn(parameters.testGroup, 'innerloop', 'clrinterpreter') }}:
         - '${{ parameters.runtimeFlavor }}_common_test_build_p1_AnyOS_AnyCPU_${{parameters.buildConfig }}'
       - ${{ if ne(parameters.stagedBuild, true) }}:
-        - ${{ format('{0}_{1}_product_build_{2}{3}_{4}_{5}', parameters.runtimeFlavor, parameters.runtimeVariant, parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
+        - ${{ if or( eq(parameters.runtimeVariant, 'minijit'),   eq(parameters.runtimeVariant, 'monointerpreter'))  }}:
```
Comment (safern) [inline eng/pipelines/common/templates/runtimes/run-test-job.yml:66]: Nit: extra spaces.

```suggestion
        - ${{ if or( eq(parameters.runtimeVariant, 'minijit'), eq(parameters.runtimeVariant, 'monointerpreter'))  }}: 
```

File: eng/pipelines/common/templates/runtimes/run-test-job.yml [hunk 4]
```diff
@@ -63,7 +63,11 @@ jobs:
       - ${{ if notIn(parameters.testGroup, 'innerloop', 'clrinterpreter') }}:
         - '${{ parameters.runtimeFlavor }}_common_test_build_p1_AnyOS_AnyCPU_${{parameters.buildConfig }}'
       - ${{ if ne(parameters.stagedBuild, true) }}:
-        - ${{ format('{0}_{1}_product_build_{2}{3}_{4}_{5}', parameters.runtimeFlavor, parameters.runtimeVariant, parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
+        - ${{ if or( eq(parameters.runtimeVariant, 'minijit'),   eq(parameters.runtimeVariant, 'monointerpreter'))  }}: 
+          # minijit and mono interpreter runtimevariants do not require any special build of the runtime
+          - ${{ format('{0}_{1}_product_build_{2}{3}_{4}_{5}', parameters.runtimeFlavor, '', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
+        - ${{ if not(or(eq(parameters.runtimeVariant, 'minijit'),   eq(parameters.runtimeVariant, 'monointerpreter')))  }}:
```
Comment (safern) [inline eng/pipelines/common/templates/runtimes/run-test-job.yml:69]:
```suggestion
        - ${{ if not(or(eq(parameters.runtimeVariant, 'minijit'), eq(parameters.runtimeVariant, 'monointerpreter')))  }}:
```

File: src/tests/issues.targets (hunk 2 above)
[the remark this answers is not in the document]
Comment (safern) [reply src/tests/issues.targets:2676]: If we don't need it for mono might be worth removing to keep the condition "simpler".

[the remark this answers is not in the document]
Comment (fanyang-mono) [reply src/tests/issues.targets:2676]: > If we don't need it for mono might be worth removing to keep the condition "simpler".

I agree.

[the remark this answers is not in the document]
Comment (trylek) [reply src/tests/issues.targets:2676]: I believe that AltJit is / was internally used by the JIT team to test and develop cross-compilation; in this mode, there are two JITs involved - first the code is run through the AltJit and then it's "actually" JITted using the "normal" JIT. In any case I don't see why we'd need it for Mono runtime testing.

**Review by trylek — APPROVED** [review]
LGTM, thank you!

File: eng/pipelines/common/templates/runtimes/run-test-job.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (naricc) [reply eng/pipelines/common/templates/runtimes/run-test-job.yml:66]: If unassigned, variables in in these azure pipelines ymls have the default value '' (the empty string).

File: src/tests/issues.targets (hunk 2 above)
[the remark this answers is not in the document]
Comment (naricc) [reply src/tests/issues.targets:2676]: I will remove it.

**Comment by imhameed** [comment]
`_il_relu_conv`, `deadoponerrorinfunclet_r`, and `deadoponerrorinfunclet_d` should be disabled for LLVM AOT; these are genuine failures.

**Comment by naricc** [comment]
Looks like hte libraries failures are this issue: https://github.com/dotnet/runtime/issues/52031

**Comment by naricc** [comment]
Also this one: https://github.com/dotnet/core-eng/issues/12969

**Comment by naricc** [comment]
Some how I didn't merge this. Re-opening.

**Comment by SamMonoRT** [comment]
Don't see how your changes can cause Library/WASM lane failures @steveisok are these known issues and we can merge this PR ?

**Pull request merged**